numpy/core/src/multiarray/alloc.c

```diff
@@ -470,50 +470,80 @@ PyDataMem_UserRENEW(void *ptr, size_t size, const PyDataMemAllocator *allocator)
 }
 
 /*NUMPY_API
- * Sets a new allocation policy. If the input value is NULL, will reset
- * the policy to the default. Returns the previous policy. We wrap
- * the user-provided functions so they will still call the python
- * and numpy memory management callback hooks.
+ * Set a new allocation policy. If the input value is NULL, will reset
+ * the policy to the default. Return the previous policy, or set an exception
+ * and return NULL if an error has occurred. We wrap the user-provided
+ * functions so they will still call the python and numpy
+ * memory management callback hooks.
  */
 NPY_NO_EXPORT const PyDataMem_Handler *
 PyDataMem_SetHandler(PyDataMem_Handler *handler)
 {
-    const PyDataMem_Handler *old = current_handler;
-    if (handler) {
-        current_handler = handler;
+    PyObject *capsule;
+    PyObject *old_capsule;
+    PyDataMem_Handler *old_handler;
+    PyObject *token;
+    if (PyContextVar_Get(current_handler, NULL, &old_capsule)) {
+        return NULL;
+    }
+    old_handler = (PyDataMem_Handler *) PyCapsule_GetPointer(old_capsule, NULL);
+    Py_DECREF(old_capsule);
+    if (old_handler == NULL) {
+        return NULL;
+    }
+    if (handler != NULL) {
+        capsule = PyCapsule_New(handler, NULL, NULL);
+        if (capsule == NULL) {
+            return NULL;
+        }
     }
     else {
-        current_handler = &default_handler;
+        capsule = PyCapsule_New(&default_handler, NULL, NULL);
+        if (capsule == NULL) {
+            return NULL;
+        }
+    }
+    token = PyContextVar_Set(current_handler, capsule);
+    Py_DECREF(capsule);
+    if (token == NULL) {
+        return NULL;
     }
-    return old;
+    Py_DECREF(token);
+    return old_handler;
 }
 
 /*NUMPY_API
- * Return the const char name of the PyDataMem_Handler used by the
- * PyArrayObject or its base. If neither the PyArrayObject owns its own data
- * nor its base is a PyArrayObject which owns its own data return an empty string.
- * If NULL, return the name of the current global policy that
- * will be used to allocate data for the next PyArrayObject.
+ * Return the PyDataMem_Handler used by the PyArrayObject. If NULL, return
+ * the current global policy that ill be used to allocate data
+ * for the next PyArrayObject. On failure, set an exception and return NULL.
  */
-NPY_NO_EXPORT const char *
-PyDataMem_GetHandlerName(PyArrayObject *obj)
+NPY_NO_EXPORT const PyDataMem_Handler *
+PyDataMem_GetHandler(PyArrayObject *obj)
 {
+    PyObject *base;
+    PyObject *capsule;
+    PyDataMem_Handler *handler;
     if (obj == NULL) {
-        return current_handler->name;
+        if (PyContextVar_Get(current_handler, NULL, &capsule)) {
```

Comment:
This can set `capsule=NULL` based on my reading of the docs, in which case this should probably error.

Reply:
In this case, the context var has a `default_value`, so it will always return something.
https://github.com/eliaskoromilas/numpy/blob/configurable_allocator/numpy/core/src/multiarray/multiarraymodule.c#L4918

Reply:
Another concern; does `PyContextVar_New` require the GIL to be held?

Reply:
Not to my knowledge.

https://github.com/python/cpython/blob/main/Python/context.c#L806-L842